sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import com.rabbitmq.client.AlreadyClosedException;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.MissedHeartbeatException;
+import com.rabbitmq.client.PossibleAuthenticationFailureException;
+import com.rabbitmq.client.ShutdownSignalException;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ProtocolException;
+import java.net.URISyntaxException;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Throwables;
+
+/**
+ * RabbitMQ multiplexes over a single Connection using Channels so it should not be necessary to
+ * open multiple Connections to a single host. This class implements {@link ChannelLeaser} by
+ * maintaining a map of lessees and their associated Channels over a single Connection.
+ *
+ * <p>Note: it is unclear what, if anything, is ultimately responsible for calling close() here; it
+ * should only happen on Beam runner shutdown, or if it were knowable that there were no current
+ * Readers or Writers interacting with Rabbit.
+ */
+@ThreadSafe
+class ConnectionHandler implements ChannelLeaser, Closeable {
+  private final Map<UUID, Channel> channelsByLessee = new ConcurrentHashMap<>();
+  private final String uri;
+  private volatile Connection connection;
+
+  public ConnectionHandler(String uri) {
+    this.uri = uri;
+  }
+
+  @Override
+  public <T> T useChannel(UUID lesseeId, ChannelLeaser.UseChannelFunction<T> f) throws IOException {
```

Comment:
That's the crux of it. It multiple objects need to utilize the same Channel (such as a CheckpointMark and a Reader), I want to ensure that the Channel is shut down when appropriate. If both CheckpointMark and Reader held a reference to a `Channel` it would be difficult to safely re-establish a Channel or Connection if a recoverable error occurs or consistently close now-unusable resources of an unrecoverable error occurs. I was aiming to move the responsibility of managing the persistent global state of these resources outside of the readers and writers.